components/clm/src/biogeochem/LSparseMatMod.F90

```diff
@@ -0,0 +1,448 @@
+
+  allocate(next)
+
+  next%val=val
+  next%icol=icol
+  next%irow=irow
+  next%next=> self
+  self => next
+  nelms=nelms+1
+  end subroutine spm_list_insert
+!---------------------------------------------------------------
+
+  function spm_list_next(self)result(next)
+
+  implicit none
+  type(spm_list_type), pointer :: self
+  type(spm_list_type), pointer :: next
+
+  next => self%next
+  end function spm_list_next
+!---------------------------------------------------------------
+
+  subroutine spm_list_free(self)
+  !
+  ! DESCRIPTION
+  ! Free the allocated memory associated with the sparse matrix
+  implicit none
+  type(spm_list_type), pointer :: self
+  type(spm_list_type), pointer :: current
+  type(spm_list_type), pointer :: elem
+
+  elem => self
+  do while(associated(elem))
+    current => elem
+    elem => current%next
+    deallocate(current)
+  enddo
+  end subroutine spm_list_free
+!---------------------------------------------------------------
+
+  subroutine spm_list_to_mat(spm_list, spm, nelms, szcol)
+  !
+  !copy the list to sparse matrix, and free the list
+  implicit none
+  type(spm_list_type), pointer :: spm_list
+  class(sparseMat_type), pointer :: spm
+  integer, intent(in) :: nelms    !total number of nonzero elements
+  integer, intent(in) :: szcol    !total columns of the matrix
+
+  type(spm_list_type), pointer :: next
+  integer :: nrows
+  integer :: irow, jj
+
+  nrows=spm_list%irow
+
+  allocate(spm, source=create_spm_type())
+
+  call spm%init(nelms, nrows, szcol)
+
+  jj = nelms
+  irow=nrows
+  next => spm_list
+  do while(associated(next))
+    if (irow /= next%irow)then
+       spm%pB(irow)=jj+1
+       irow=next%irow
+    endif
+    spm%val(jj) = next%val
+    spm%icol(jj)= next%icol
+    spm%ncol(irow)=spm%ncol(irow)+1
+    next=>spm_list_next(next)
+    jj=jj-1
+  enddo
+  !fill the last row
+  spm%pB(irow)=jj+1
+  call spm_list_free(spm_list)
+  end subroutine spm_list_to_mat
+
+
+
+  !-------------------------------------------------------------------------------
+  subroutine calc_state_pscal(this, nprimvars, dtime, ystate, p_dt,  d_dt, pscal, lneg, errinfo)
+    !
+    ! !DESCRIPTION:
+    ! calculate limiting factor from each primary state variable
+    !
+    use BetrstatusType     , only : betr_status_type
+    use betr_constants     , only : betr_errmsg_len
+    implicit none
+    ! !ARGUMENTS:
+    class(lom_type), intent(in) :: this
+    integer,  intent(in)  :: nprimvars
+    real(r8), intent(in)  :: dtime
+    real(r8), intent(in)  :: ystate(1:nprimvars)
+    real(r8), intent(in)  :: p_dt(1:nprimvars)
+    real(r8), intent(in)  :: d_dt(1:nprimvars)
+    real(r8), intent(out) :: pscal(1:nprimvars)
+    logical,  intent(out) :: lneg
+    integer,  intent(out) :: errinfo
+    character(len=betr_errmsg_len) :: msg
+
+    ! !LOCAL VARIABLES:
+    real(r8) :: yt
+    integer  :: j
+    real(r8),parameter :: p_par=0.999_r8
+    real(r8), parameter :: tiny_val=-1.e-14_r8
```

Comment:
@jinyuntang, Should the definition of `tiny_val` within this subroutine be omitted because it is also defined for the module at [Line-34](https://github.com/E3SM-Project/E3SM/pull/2946/files#diff-6c09cf12ed23f28ed428bd168b62a143R34)?

Reply:
@bishtgautam, Thanks for catching this, I just removed this duplicated line.